core/src/impl/Kokkos_BitOps.hpp

```diff
@@ -57,93 +57,202 @@ namespace Kokkos {
   return full != i ? __builtin_ffs(~i) - 1 : -1;
 #else
+  return bit_first_zero_fallback(i);
+#endif
+}
+
+#if defined(__EDG__)
+#pragma push
+#pragma diag_suppress implicit_return_from_non_void_function
+#endif
+KOKKOS_FORCEINLINE_FUNCTION
+int bit_first_zero(unsigned i) noexcept {
+  KOKKOS_IF_ON_DEVICE((return bit_first_zero_device(i);))
+  KOKKOS_IF_ON_HOST((return bit_first_zero_host(i);))
+}
+#if defined(__EDG__)
+#pragma pop
+#endif
+
+KOKKOS_FORCEINLINE_FUNCTION
+int bit_scan_forward_fallback(unsigned i) {
   int offset = -1;
-  if (full != i) {
-    for (offset = 0; i & (1 << offset); ++offset)
+  if (i) {
+    for (offset = 0; (i & (1 << offset)) == 0; ++offset)
       ;
   }
   return offset;
-#endif
 }
 
-KOKKOS_FORCEINLINE_FUNCTION
-int bit_scan_forward(unsigned i) {
-#if defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__)
+KOKKOS_IMPL_DEVICE_FUNCTION inline int bit_scan_forward_device(unsigned i) {
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP)
   return __ffs(i) - 1;
 #elif defined(KOKKOS_COMPILER_INTEL)
   return _bit_scan_forward(i);
+#else
+  return bit_scan_forward_fallback(i);
+#endif
+}
+
+KOKKOS_IMPL_HOST_FUNCTION inline int bit_scan_forward_host(unsigned i) {
+#if defined(KOKKOS_COMPILER_INTEL)
+  return _bit_scan_forward(i);
 #elif defined(KOKKOS_COMPILER_CRAYC)
   return i ? _popcnt(~i & (i - 1)) : -1;
 #elif defined(KOKKOS_COMPILER_GNU) || defined(__GNUC__) || defined(__GNUG__)
   return __builtin_ffs(i) - 1;
 #else
-  int offset = -1;
-  if (i) {
-    for (offset = 0; (i & (1 << offset)) == 0; ++offset)
-      ;
-  }
-  return offset;
+  return bit_scan_forward_fallback(i);
+#endif
+}
+
+#if defined(__EDG__)
+#pragma push
+#pragma diag_suppress implicit_return_from_non_void_function
 #endif
+KOKKOS_FORCEINLINE_FUNCTION
+int bit_scan_forward(unsigned i) {
+  KOKKOS_IF_ON_DEVICE((return bit_scan_forward_device(i);))
+  KOKKOS_IF_ON_HOST((return bit_scan_forward_host(i);))
 }
+#if defined(__EDG__)
+#pragma pop
+#endif
 
 /// Count the number of bits set.
 KOKKOS_FORCEINLINE_FUNCTION
-int bit_count(unsigned i) {
-#if defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__)
+int bit_count_fallback(unsigned i) {
+  // http://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetNaive
+  i = i - ((i >> 1) & ~0u / 3u);                           // temp
+  i = (i & ~0u / 15u * 3u) + ((i >> 2) & ~0u / 15u * 3u);  // temp
+  i = (i + (i >> 4)) & ~0u / 255u * 15u;                   // temp
+
+  // count
+  return (int)((i * (~0u / 255u)) >> (sizeof(unsigned) - 1) * CHAR_BIT);
+}
+
+KOKKOS_IMPL_DEVICE_FUNCTION inline int bit_count_device(unsigned i) {
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP)
   return __popc(i);
 #elif defined(__INTEL_COMPILER)
   return _popcnt32(i);
+#else
+  return bit_count_fallback(i);
+#endif
+}
+
+KOKKOS_IMPL_HOST_FUNCTION inline int bit_count_host(unsigned i) {
+#if defined(__INTEL_COMPILER)
```

Comment:
Note this is using the `__INTEL_COMPILER` macro directly

Reply:
I can confirm that `icpx` is fine with `_popcnt32` even on the device so using `KOKKOS_COMPILER_INTEL`  should be fine here (and then also using on the device).
I'm happy to create a follow-up pull request, if you don't want to change it here.